### ISSUE TITLE
parser: v4.1+ breaks an eslint-plugin-react `no-unused-vars` test that passes on v4.0

### ISSUE THREAD

**Issue by ljharb** [typescript-eslint/typescript-eslint#3788]
Specifically, https://github.com/ljharb/eslint-plugin-react/runs/3415051454 shows the failures, but when I pin it to v4.0, they pass: (link TBD)

This seems like a breaking change, but if there's a simple config change I can do to make the tests pass (and to tell users if they complain) that would likely suffice.

**Comment by bradzacher** [comment]
v4.0 introduced our new scope analyser, and v4.1 added support for handling the JSX pragma reference (pr: #2498).

The reason your test is failing is because it's expecting there to be a `no-unused-vars` error, but there's not because our scope analyser creates a reference to `React` which `no-unused-vars` is able to understand.

No users should complain unless for some weird reason they want to see unused variable errors against `React`.

There are config options you can provide to control this behaviour:
- https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/parser#parseroptionsjsxpragma
- https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/parser#parseroptionsjsxfragmentname

**Comment by ljharb** [comment]
Hmm, so you're saying that the TS parser as of v4.1 *automatically* adds a React import even if it's not present in the code??

**Comment by bradzacher** [comment]
To clarify:

As of 4.1 our scope analyser just understands that

```ts
var React;
<></>
```

transpiles to
```ts
var React;
React.jsx(React.Fragment,{});
```

So the scope analyser creates a reference at the JSX location to the top-level `React` variable.

Example:

This code fragment:
https://github.com/typescript-eslint/typescript-eslint/blob/043462ef4d8f3aa16c10efb84fea25a193deefac/packages/scope-manager/tests/fixtures/jsx/factory/default-jsxPragma-fragment.tsx#L3-L5

Creates a scope tree with the `React` variable (note the `references` array has one reference in it):
https://github.com/typescript-eslint/typescript-eslint/blob/043462ef4d8f3aa16c10efb84fea25a193deefac/packages/scope-manager/tests/fixtures/jsx/factory/default-jsxPragma-fragment.tsx.shot#L7-L27

---

Put another way - if they are using at least v4.1 of our parser - users do not need the `react/jsx-uses-react` rule as it will do nothing.

**Comment by ljharb** [comment]
ah, gotcha, thanks - that clarifies it.

That's a bit unfortunate in one way - I'm encouraging people to extend `react/jsx-runtime` when using the new JSX runtime, and that won't configure the TS resolver properly.

**Comment by bradzacher** [comment]
No reason why you couldn't add `parserOptions: { jsxPragma: null }` to the config!
All parsers (that I know of) will ignore that option - so it'll just impact people using our parser.

**Comment by ljharb** [comment]
hmm, that doesn't seem to be making the test pass - is there a way i can modify the test case so the TS eslint parser *doesn't* mark `React` as used for me, since that's the entire purpose of react/jsx-uses-react?

**Comment by smashercosmo** [comment]
Experiencing the same issue. Getting false negatives from no-unused-vars rule. I'm not using react's automatic runtime, but typescript-eslint reports React as unused variable. Related https://github.com/typescript-eslint/typescript-eslint/issues/3303

**Comment by ljharb** [comment]
@bradzacher ping on the above question :-)

**Comment by bradzacher** [comment]
It should be `jsxPragma: null` - that will tell our scope analyser to not automatically reference `React`

https://github.com/typescript-eslint/typescript-eslint/blob/02c6ff3c5a558f9308d7166d524156dc12e32759/packages/eslint-plugin/tests/rules/no-unused-vars/no-unused-vars.test.ts#L1580-L1605

---

@smashercosmo - please don't piggy back on issues - file a new issue and follow the issue template to get support.
